File: EmuSen/Cores/Nintendo/Mercury - GB-GBC/Memory/Cartridge.cs
using System;
using System.IO;
using System.Text;
using EmuSen.Common;
using EmuSen.Cores.Nintendo.Mercury.Memory.Mappers;
using EmuSen.Galaxia.Library;

namespace EmuSen.Cores.Nintendo.Mercury.Memory
{
    // What the $0143 flag says about colour support - see Mercury_Memory.md §2.1.
    public enum CgbSupport
    {
        None,
        Enhanced,
        Required,
    }

    // One Game Boy image and the board it describes - see Mercury_Memory.md §2.
    public sealed class Cartridge
    {
        public const int RomBankSize = 0x4000;
        public const int RamBankSize = 0x2000;

        public const int HeaderStart = 0x0134;
        public const int HeaderChecksumAddress = 0x014D;

        [SkipInState] public byte[] Rom = Array.Empty<byte>();

        public byte[] Ram = Array.Empty<byte>();

        [SkipInState] public string Title = "";
        [SkipInState] public byte CartridgeType;
        [SkipInState] public CgbSupport Cgb;
        [SkipInState] public bool SuperGameBoy;
        [SkipInState] public bool HasBattery;
        [SkipInState] public bool HasTimer;
        [SkipInState] public bool HasRumble;
        [SkipInState] public bool HeaderChecksumValid;
        [SkipInState] public string RomPath = "";

        [SkipInState] public IMapper Mapper = null!;

        private string? _savePath;

        public int RomBanks => Rom.Length / RomBankSize;
        public int RamBanks => Ram.Length / RamBankSize;

        public static Cartridge Load(string path)
        {
            var cart = FromImage(File.ReadAllBytes(path));
            cart.RomPath = path;
            cart._savePath = Path.ChangeExtension(path, ".srm");
            cart.LoadSram();
            return cart;
        }

        public static Cartridge FromImage(byte[] image)
        {
            if (image.Length < 0x0150)
            {
                throw new InvalidDataException($"Not a Game Boy image: {image.Length} bytes is shorter than the 336-byte header.");
            }

            var cart = new Cartridge { Rom = image };

            cart.CartridgeType = image[0x0147];
            cart.Cgb = image[0x0143] switch
            {
                0xC0 => CgbSupport.Required,
                0x80 => CgbSupport.Enhanced,
                _ => CgbSupport.None,
            };
            cart.SuperGameBoy = image[0x0146] == 0x03;
            cart.Title = ReadTitle(image, cart.Cgb != CgbSupport.None);

            DescribeType(cart.CartridgeType, out bool battery, out bool timer, out bool rumble);
            cart.HasBattery = battery;
            cart.HasTimer = timer;
            cart.HasRumble = rumble;

            cart.Ram = new byte[RamSizeFor(image[0x0149], cart.CartridgeType)];
            cart.HeaderChecksumValid = ComputeHeaderChecksum(image) == image[HeaderChecksumAddress];

            // The header's own size byte is advisory; the file is the authority - see Mercury_Memory.md §2.2.
            cart.Mapper = BuildMapper(cart);
            return cart;
        }

        // Trailing NULs and, on a colour cart, the manufacturer/CGB bytes are not part of the name.
        private static string ReadTitle(byte[] image, bool isCgb)
        {
            int length = isCgb ? 11 : 16;
            var text = new StringBuilder(length);

            for (int i = 0; i < length; i++)
            {
                byte b = image[HeaderStart + i];
                if (b == 0x00) break;
                text.Append(b >= 0x20 && b < 0x7F ? (char)b : ' ');
            }

            return text.ToString().TrimEnd();
        }

        // x = x - byte - 1 across $0134-$014C; a real boot ROM locks up when this disagrees.
        public static byte ComputeHeaderChecksum(byte[] image)
        {
            byte checksum = 0;
            for (int address = HeaderStart; address <= 0x014C; address++)
            {
                checksum = (byte)(checksum - image[address] - 1);
            }
            return checksum;
        }

        private static void DescribeType(byte type, out bool battery, out bool timer, out bool rumble)
        {
            battery = type is 0x03 or 0x06 or 0x09 or 0x0D or 0x0F or 0x10 or 0x13 or 0x1B or 0x1E or 0x22 or 0xFF;
            timer = type is 0x0F or 0x10;
            rumble = type is 0x1C or 0x1D or 0x1E or 0x22;
        }

        // MBC2 carries 512 half-bytes on the chip itself and reports no size in the header.
        private static int RamSizeFor(byte code, byte cartridgeType)
        {
            if (cartridgeType is 0x05 or 0x06) return 512;

            return code switch
            {
                0x02 => 8 * 1024,
                0x03 => 32 * 1024,
                0x04 => 128 * 1024,
                0x05 => 64 * 1024,
                _ => 0,
            };
        }

        private static IMapper BuildMapper(Cartridge cart) => cart.CartridgeType switch
        {
            0x00 or 0x08 or 0x09 => new NoMbc(cart),
            >= 0x01 and <= 0x03 => new Mbc1(cart),
            0x05 or 0x06 => new Mbc2(cart),
            >= 0x0F and <= 0x13 => new Mbc3(cart),
            >= 0x19 and <= 0x1E => new Mbc5(cart),
            var other => throw new NotSupportedException(
                $"Cartridge type ${other:X2} is not implemented - see Mercury_Memory.md §4."),
        };

        private void LoadSram()
        {
            if (!HasBattery || _savePath is null || Ram.Length == 0) return;
            if (AtomicFile.TryRead(_savePath) is not { } saved) return;

            Array.Copy(saved, Ram, Math.Min(saved.Length, Ram.Length));
        }

        public void SaveSram()
        {
            if (!HasBattery || _savePath is null || Ram.Length == 0) return;
            AtomicFile.Write(_savePath, Ram);
        }
    }
}

File: EmuSen/Cores/Nintendo/MercuryRT - GB/src/memory/cartridge.rs
//! C#'s `Memory/Cartridge.cs`: one image and the board its header describes. See Mercury_Memory.md §2.

use crate::Skip;
use crate::memory::mappers::Mapper;
use crate::state::{StateReader, StateResult, StateWriter};

pub const ROM_BANK_SIZE: usize = 0x4000;
pub const RAM_BANK_SIZE: usize = 0x2000;
pub const HEADER_START: usize = 0x0134;
pub const HEADER_CHECKSUM_ADDRESS: usize = 0x014D;

/// What the `$0143` flag says about colour support.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub enum CgbSupport {
    #[default]
    None,
    Enhanced,
    Required,
}

/// Why an image is refused, as C#'s `Cartridge.FromImage` refuses it.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum RomError {
    TooShort(usize),
    UnsupportedType(u8),
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct Cartridge {
    pub rom: Skip<Vec<u8>>,
    pub ram: Vec<u8>,
    pub title: Skip<String>,
    pub cartridge_type: Skip<u8>,
    pub cgb: Skip<CgbSupport>,
    pub super_game_boy: Skip<bool>,
    pub has_battery: Skip<bool>,
    pub has_timer: Skip<bool>,
    pub has_rumble: Skip<bool>,
    pub header_checksum_valid: Skip<bool>,
    /// C#'s `_savePath`: null until a load or a battery save names it, then a string (Mercury_Native.md §6.1, D3).
    pub save_path: Option<String>,
}

impl Cartridge {
    /// `Cartridge.FromImage`, with the board it builds; the save path is the host's, as `LoadSram` sets it.
    pub fn from_image(image: Vec<u8>, save_path: Option<String>) -> Result<(Cartridge, Mapper), RomError> {
        if image.len() < 0x0150 {
            return Err(RomError::TooShort(image.len()));
        }
        let kind = image[0x0147];
        let cgb = match image[0x0143] {
            0xC0 => CgbSupport::Required,
            0x80 => CgbSupport::Enhanced,
            _ => CgbSupport::None,
        };
        let mapper = Mapper::for_type(kind).ok_or(RomError::UnsupportedType(kind))?;
        let cart = Cartridge {
            title: Skip(read_title(&image, cgb != CgbSupport::None)),
            cartridge_type: Skip(kind),
            cgb: Skip(cgb),
            super_game_boy: Skip(image[0x0146] == 0x03),
            has_battery: Skip(matches!(kind, 0x03 | 0x06 | 0x09 | 0x0D | 0x0F | 0x10 | 0x13 | 0x1B | 0x1E | 0x22 | 0xFF)),
            has_timer: Skip(matches!(kind, 0x0F | 0x10)),
            has_rumble: Skip(matches!(kind, 0x1C | 0x1D | 0x1E | 0x22)),
            ram: vec![0; ram_size_for(image[0x0149], kind)],
            header_checksum_valid: Skip(header_checksum(&image) == image[HEADER_CHECKSUM_ADDRESS]),
            save_path,
            rom: Skip(image),
        };
        Ok((cart, mapper))
    }

    pub fn rom_banks(&self) -> usize {
        self.rom.len() / ROM_BANK_SIZE
    }

    pub fn is_cgb(&self) -> bool {
        *self.cgb != CgbSupport::None
    }

    pub fn write_state(&self, w: &mut StateWriter) {
        w.bytes("Ram", &self.ram);
        w.string("_savePath", self.save_path.as_deref());
    }

    pub fn read_state(&mut self, r: &mut StateReader) -> StateResult {
        r.bytes(&mut self.ram)?; // Ram
        self.save_path = Some(r.string()?); // _savePath
        Ok(())
    }

    /// A class-typed `_cart` field elsewhere in the state: the present flag, then the same two fields.
    pub fn write_as_field(&self, w: &mut StateWriter, name: &str) {
        w.group_class(name, |w| self.write_state(w));
    }

    pub fn read_as_field(&mut self, r: &mut StateReader) -> StateResult {
        if r.bool()? {
            self.read_state(r)?;
        }
        Ok(())
    }
}

/// Trailing NULs and, on a colour cart, the manufacturer and CGB bytes are not part of the name.
fn read_title(image: &[u8], cgb: bool) -> String {
    let length = if cgb { 11 } else { 16 };
    let mut text = String::new();
    for &b in &image[HEADER_START..HEADER_START + length] {
        if b == 0 {
            break;
        }
        text.push(if (0x20..0x7F).contains(&b) { b as char } else { ' ' });
    }
    text.trim_end().to_string()
}

/// `x = x - byte - 1` across `$0134-$014C`.
pub fn header_checksum(image: &[u8]) -> u8 {
    image[HEADER_START..=0x014C].iter().fold(0u8, |sum, &b| sum.wrapping_sub(b).wrapping_sub(1))
}

/// MBC2 carries 512 half-bytes on the chip and reports no size in the header.
fn ram_size_for(code: u8, kind: u8) -> usize {
    if matches!(kind, 0x05 | 0x06) {
        return 512;
    }
    match code {
        0x02 => 8 * 1024,
        0x03 => 32 * 1024,
        0x04 => 128 * 1024,
        0x05 => 64 * 1024,
        _ => 0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn image(kind: u8, ram_code: u8, cgb: u8) -> Vec<u8> {
        let mut rom = vec![0u8; 0x8000];
        rom[0x0134..0x013B].copy_from_slice(b"WISEMAN");
        rom[0x0143] = cgb;
        rom[0x0147] = kind;
        rom[0x0149] = ram_code;
        rom[HEADER_CHECKSUM_ADDRESS] = header_checksum(&rom);
        rom
    }

    #[test]
    fn the_header_decides_the_board_the_ram_and_the_colour_mode() {
        let (cart, mapper) = Cartridge::from_image(image(0x03, 0x02, 0x00), None).unwrap();
        assert!(matches!(mapper, Mapper::Mbc1(_)));
        assert_eq!((cart.ram.len(), *cart.has_battery, cart.is_cgb(), cart.title.as_str()), (8192, true, false, "WISEMAN"));
        assert!(*cart.header_checksum_valid);
        let (cart, mapper) = Cartridge::from_image(image(0x06, 0x03, 0xC0), None).unwrap();
        assert!(matches!(mapper, Mapper::Mbc2(_)));
        assert_eq!((cart.ram.len(), *cart.cgb), (512, CgbSupport::Required));
        let (cart, mapper) = Cartridge::from_image(image(0x10, 0x03, 0x80), None).unwrap();
        assert!(matches!(mapper, Mapper::Mbc3(_)));
        assert_eq!((cart.ram.len(), *cart.has_timer, *cart.cgb), (32768, true, CgbSupport::Enhanced));
        let (cart, mapper) = Cartridge::from_image(image(0x1E, 0x04, 0x00), None).unwrap();
        assert!(matches!(mapper, Mapper::Mbc5(_)));
        assert_eq!((cart.ram.len(), *cart.has_rumble, *cart.has_battery), (131072, true, true));
        assert!(matches!(Cartridge::from_image(image(0x09, 0x05, 0x00), None).unwrap().1, Mapper::NoMbc(_)));
    }

    #[test]
    fn what_csharp_refuses_is_refused() {
        assert_eq!(Cartridge::from_image(vec![0; 0x14F], None).err(), Some(RomError::TooShort(0x14F)));
        assert_eq!(Cartridge::from_image(image(0x22, 0, 0), None).err(), Some(RomError::UnsupportedType(0x22)));
        assert_eq!(Cartridge::from_image(image(0x04, 0, 0), None).err(), Some(RomError::UnsupportedType(0x04)));
    }
}
